In [29]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [30]:
df= pd.read_csv("/content/heart.csv")

In [31]:
df

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

In [32]:
df= df.drop(['oldpeak','slp','thall'],axis=1)

# Understanding our DataSet:

#### Age : Age of the patient

#### Sex : Sex of the patient

#### exang: exercise induced angina (1 = yes; 0 = no)

#### ca: number of major vessels (0-3)

#### cp : Chest Pain type chest pain type

- Value 0: typical angina
- Value 1: atypical angina
- Value 2: non-anginal pain
- Value 3: asymptomatic

#### trtbps : resting blood pressure (in mm Hg)

#### chol : cholestoral in mg/dl fetched via BMI sensor

#### fbs : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)

#### rest_ecg : resting electrocardiographic results

- Value 0: normal
- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

#### thalach : maximum heart rate achieved

#### target : 0= less chance of heart attack 1= more chance of heart attack

In [33]:
df

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  caa  output
0     63    1   3     145   233    1        0       150     0    0       1
1     37    1   2     130   250    0        1       187     0    0       1
2     41    0   1     130   204    0        0       172     0    0       1
3     56    1   1     120   236    0        1       178     0    0       1
4     57    0   0     120   354    0        1       163     1    0       1
..   ...  ...  ..     ...   ...  ...      ...       ...   ...  ...     ...
298   57    0   0     140   241    0        1       123     1    0       0
299   45    1   3     110   264    0        1       132     0    0       0
300   68    1   0     144   193    1        1       141     0    2       0
301   57    1   0     130   131    0        1       115     1    1       0
302   57    0   1     130   236    0        0       174     0    1       0

[303 rows x 11 columns]

In [34]:
df.shape

(303, 11)

In [35]:
df.isnull().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
caa         0
output      0
dtype: int64

In [36]:
df.corr()

age       sex        cp    trtbps      chol       fbs  \
age       1.000000 -0.098447 -0.068653  0.279351  0.213678  0.121308   
sex      -0.098447  1.000000 -0.049353 -0.056769 -0.197912  0.045032   
cp       -0.068653 -0.049353  1.000000  0.047608 -0.076904  0.094444   
trtbps    0.279351 -0.056769  0.047608  1.000000  0.123174  0.177531   
chol      0.213678 -0.197912 -0.076904  0.123174  1.000000  0.013294   
fbs       0.121308  0.045032  0.094444  0.177531  0.013294  1.000000   
restecg  -0.116211 -0.058196  0.044421 -0.114103 -0.151040 -0.084189   
thalachh -0.398522 -0.044020  0.295762 -0.046698 -0.009940 -0.008567   
exng      0.096801  0.141664 -0.394280  0.067616  0.067023  0.025665   
caa       0.276326  0.118261 -0.181053  0.101389  0.070511  0.137979   
output   -0.225439 -0.280937  0.433798 -0.144931 -0.085239 -0.028046   

           restecg  thalachh      exng       caa    output  
age      -0.116211 -0.398522  0.096801  0.276326 -0.225439  
sex      -0.058196 -0.044020  0.141664  0.118261 -0.280937  
cp        0.044421  0.295762 -0.394280 -0.181053  0.433798  
trtbps   -0.114103 -0.046698  0.067616  0.101389 -0.144931  
chol     -0.151040 -0.009940  0.067023  0.070511 -0.085239  
fbs      -0.084189 -0.008567  0.025665  0.137979 -0.028046  
restecg   1.000000  0.044123 -0.070733 -0.072042  0.137230  
thalachh  0.044123  1.000000 -0.378812 -0.213177  0.421741  
exng     -0.070733 -0.378812  1.000000  0.115739 -0.436757  
caa      -0.072042 -0.213177  0.115739  1.000000 -0.391724  
output    0.137230  0.421741 -0.436757 -0.391724  1.000000

In [37]:
x= df.iloc[:,:-1]
x

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  caa
0     63    1   3     145   233    1        0       150     0    0
1     37    1   2     130   250    0        1       187     0    0
2     41    0   1     130   204    0        0       172     0    0
3     56    1   1     120   236    0        1       178     0    0
4     57    0   0     120   354    0        1       163     1    0
..   ...  ...  ..     ...   ...  ...      ...       ...   ...  ...
298   57    0   0     140   241    0        1       123     1    0
299   45    1   3     110   264    0        1       132     0    0
300   68    1   0     144   193    1        1       141     0    2
301   57    1   0     130   131    0        1       115     1    1
302   57    0   1     130   236    0        0       174     0    1

[303 rows x 10 columns]

In [38]:
y= df.iloc[:,-1:]
y

output
0         1
1         1
2         1
3         1
4         1
..      ...
298       0
299       0
300       0
301       0
302       0

[303 rows x 1 columns]

In [39]:
# y= df.pop('output')
# y

In [40]:
# x = df
# x

In [41]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=101)

In [42]:
from sklearn.linear_model import LogisticRegression

In [43]:
from sklearn.preprocessing import LabelEncoder

In [44]:
logreg= LogisticRegression()

In [45]:
feature_names = ['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh', 'exng', 'caa']

In [46]:
logreg = LogisticRegression()
logreg.fit(x_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [47]:
y_pred = logreg.predict(x_test)

In [48]:
y_pred

array([0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1,
       0, 1, 0])

In [49]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [50]:
# Y_pred1 = logreg.predict(x_test)
lr_conf_matrix = confusion_matrix(y_test,y_pred )
lr_acc_score = accuracy_score(y_test, y_pred)

In [51]:
lr_conf_matrix

array([[35,  9],
       [ 5, 42]])

In [52]:
print(lr_acc_score*100,"%")

84.61538461538461 %


In [53]:
import joblib

In [54]:
joblib.dump(logreg, 'ash_model_1.pkl')

['ash_model_1.pkl']

In [55]:
joblib.dump(logreg, 'lr_heart.h5')

['lr_heart.h5']

In [57]:
# Load the model
loaded_model = joblib.load('lr_heart.h5')

# Simulate new customer data
new_customer_data = pd.DataFrame({
    'age': [30],
    'sex': [12],
    'cp': [50],
    'trtbps': [100],
    'chol': [1],
    'fbs': [1],
    'restecg': [1],
    'thalachh': [1],
    'exng': [1],
    'caa': [1]
})

# Scale the new customer data
# new_customer_data_scaled = scaler.transform(new_customer_data)

# Make predictions
churn_prediction = loaded_model.predict(new_customer_data)
print("Churn Prediction:", churn_prediction)

Churn Prediction: [1]
